[libbeat] : add processors "decode_duration", "move_fields" (#31301)
elastic/beats: libbeat/processors/move_fields/docs/move_fields.asciidoc

[[move-fields]]
=== Move fields

++++
<titleabbrev>move-fields</titleabbrev>
++++

> The `move_fields` processor allows to rearrange event fields from a object into another.

Author: I think "moves and optionally rearranges." maybe better.
The `move_fields` also can use for add prefix to fields. How we description it?

> The `move_fields` processor allows to rearrange event fields from a object into another.

Reviewer: Edited my suggestion. Feel free to commit if you like it!

    "user_id": 100,
    "message": "i'm a message",
    "rpc": {
      "method": "a",
      "elapsed_time": 100
    }
  }
}
----

To add a prefix to the whole event:

[source,json]
----
{
  "app": { "method": "a"},
  "cost": 100
}
----

use this configuration:

[source,yaml]
----
processors:
  - move_fields:
      to: "my_prefix_"
----

then your final event will be:

[source,json]
----
{
  "my_prefix_app": { "method": "a"},
  "my_prefix_cost": 100
}
----

.Move-fields options
[options="header"]
|======
| Name                    | Required | Default                  | Description                                                                                           |
| `from`                  | no       |                          | Which field you want extract. these field will be moved into `to`. empty indicate event root.         |
| `fields`                | no       |                          | filter out which fields from `from` moving to `to`. empty list indicate all fields.                   |
| `ignore_missing`        | no       | false                    | indicate ignore not found error when extract field from `fields` list.                                |

> | `exclude`               | no       |                          | filter out which fields from `from` NOT moving to `to`.                                               |

Author: Yes. it is.

> | `to`                    | yes      |                          | These fields extract from `from` destination field prefix the `to` will base on fields root.          |

Author: The `move_fields` processor first extract sub fields from `from` object then combine `to` will be new path.
For example the origin event is:
```json
{
  "a": {
    "name": "Jon",
    "age": 22
  }
}
```

use `move_fields` plugin set `from` = `a`, `to` = `my_prefix_` then the output is:
```json
{
  "a": {},
  "my_prefix_name": "Jon",
  "my_prefix_age": 22
}
```

I know it is hard to understand. so how should we to description it?

> | `to`                    | yes      |                          | These fields extract from `from` destination field prefix the `to` will base on fields root.          |

Author: Another example it is:
set `from` = `a`, `to` = 'b.' then the output is:
```json
{
  "a": {},
  "b": {
    "name": "Jon",
    "age": 22
  }
}
```

> | `to`                    | yes      |                          | These fields extract from `from` destination field prefix the `to` will base on fields root.          |

Reviewer 2: Is this any closer?

`from`: The fields to extract. This field and any nested fields will be moved to the `to` field unless they are filtered out. If `from` is empty, the event root is used. 

`to`: The destination of the extracted fields in the fields root. Added as a prefix to the fields extracted from `from`.

Something that's not clear to me in the two examples is why when `to` = `my_prefix_`, `my_prefix_` is prepended to the key values, but when `to` = `b`, the `name` and `age` fields are added as keys to the `b` object.

> | `to`                    | yes      |                          | These fields extract from `from` destination field prefix the `to` will base on fields root.          |

Reviewer: Thanks Brandon! I think the syntax is a little confusing. The period in `b.` means that the fields get added to the `b` object. That makes me wonder if `to = b.my_prefix_` would also work and what the result would be. I am struggling to come up with a good suggestion here.

> | `to`                    | yes      |                          | These fields extract from `from` destination field prefix the `to` will base on fields root.          |

Author: Yes. it's also work.
Another example below:

config:
set `form` = ``, `fields` = `other`, `to` = `app.b.my_prefix_`, `exclude` = ``.

input event: 

```json
{
  "app": {
   "version": 1,
   "method": "2"
  },
  "other": 3
}
```

process output event: 

```json
{
  "app": {
   "version": 1,
   "method": "2"
   "b": {
       "my_prefix_other": 3
    }
  }
}
```

The processor can move fields to anywhere.

It any idea for the documentation?

Thanks. 😉

[[processor-move-fields]]
=== Move fields

++++
<titleabbrev>move-fields</titleabbrev>
++++

> The `move-fields` processor allow you to arrange your message fields from a object into another object.

Reviewer 3: We use the word "event" to refer to the data a Beat is ingesting/processing. Some other bits of this documentation also need this updating from "message" to "event".
```suggestion
The `move-fields` processor allow you to rearrange your event fields from a object into another object.
```

The processor first extract fields from `parent_path` from message then use `from` and `exclude` as filter to choose which fields we should move into `to` field.

For example, you got a message below:
[source,json]
----
{
  "app": {
    "method": "a",
    "elapsed_time": 100,
    "user_id": 100,
    "message": "i'm a message"
  }
}
----

Therefore, you may want split field `method` and field `elapsed_time` into another field because they are rpc fields(or some reason).

The message after we arrange. this is better to understand and structure to store and analyze in elasticsearch.

[source,json]
----
{
  "app": {
    "user_id": 100,
    "message": "i'm a message"
  },
  "app.rpc": {
    "method": "a",
    "elapsed_time": 100
  }
}
----

Or you can do more thing like this. rename field into a object.
[source,json]
----
{
  "app.method": "a",
  "app.cost": 100
}
----

transform to
[source,json]
----
{
  "app": { "method": "a", "cost": 100 }
}
----

with this config
[source,yaml]
----
processors:
  - move_fields:
      from: [ "app.method", "app.cost" ]
----

> For add what you want prefix to whole message.

Reviewer 3: Small improvement on the copy:
```suggestion
To add a prefix to the whole event:
```

[source,json]
----
{
  "app": { "method": "a"},
  "cost": 100
}
----

> config by this

Reviewer 3: ```suggestion
use this configuration
```

[source,yaml]
----
processors:
  - move_fields:
      to: "my_prefix_"
----

> then you got message

Reviewer 3: ```suggestion
then your final event will be:
```

[source,json]
----
{
  "my_prefix_app": { "method": "a"},
  "my_prefix_cost": 100
}
----

.Move-fields options
[options="header"]
|======
| Name                    | Required | Default                  | Description                                                                                           |
| `parent_path`           | no       |                          | Which field you want extract. these field will be moved into `to`. empty indicate message root.       |
| `from`                  | no       |                          | filter out which fields from `parent_path` moving to `to`.                                            |
| `ignore_from_not_found` | no       |                          | indicate ignore not found error when extract field from `from` list.                                  |
| `exclude`               | no       |                          | filter out which fields from `parent_path` NOT moving to `to`.                                        |

> | `to`                    | no       |                          | These fields extract from `parent_path` destination field prefix the `to` will base on fields root.   |

Reviewer 3: If no configuration is required, what does the default config do?

It's better to have some of them required to make sure the processor will always do something useful. Based on your examples, it seems that at least `to` should be required.
